File: go/txline/odds.go
package txline

import (
	"context"
	"net/url"
	"strconv"
)

type OddsClient struct {
	client *Client
}

func (o OddsClient) Snapshot(ctx context.Context, fixtureID int64, asOf *int64) ([]OddsPayload, error) {
	query := url.Values{}
	if asOf != nil {
		addInt64(query, "asOf", *asOf)
	}
	var out []OddsPayload
	err := o.client.getJSON(ctx, "/odds/snapshot/"+strconv.FormatInt(fixtureID, 10), query, true, &out)
	return out, err
}

func (o OddsClient) LiveUpdatesByFixture(ctx context.Context, fixtureID int64) ([]OddsPayload, error) {
	var out []OddsPayload
	err := o.client.getJSON(ctx, "/odds/updates/"+strconv.FormatInt(fixtureID, 10), nil, true, &out)
	return out, err
}

func (o OddsClient) HistoricalUpdates(ctx context.Context, epochDay uint32, hourOfDay, interval uint8, fixtureID *int64) ([]OddsPayload, error) {
	if err := validateHour(hourOfDay); err != nil {
		return nil, err
	}
	if err := validateInterval(interval); err != nil {
		return nil, err
	}
	query := url.Values{}
	if fixtureID != nil {
		addInt64(query, "fixtureId", *fixtureID)
	}
	endpoint := "/odds/updates/" + strconv.FormatUint(uint64(epochDay), 10) + "/" + strconv.Itoa(int(hourOfDay)) + "/" + strconv.Itoa(int(interval))
	var out []OddsPayload
	err := o.client.getJSON(ctx, endpoint, query, true, &out)
	return out, err
}

func (o OddsClient) Validation(ctx context.Context, messageID string, ts int64) (OddsValidation, error) {
	query := url.Values{}
	query.Set("messageId", messageID)
	addInt64(query, "ts", ts)
	var out OddsValidation
	err := o.client.getJSON(ctx, "/odds/validation", query, true, &out)
	return out, err
}

File: typescript/src/http/odds.ts
import type { TxlineClient } from "../client.js";
import { validateHour, validateInterval } from "./fixtures.js";
import type { OddsPayload, OddsValidation } from "./models.js";

export class OddsClient {
  constructor(private readonly client: TxlineClient) {}

  async snapshot(fixtureId: number, asOf?: number): Promise<OddsPayload[]> {
    return await this.client.getJson(`/odds/snapshot/${fixtureId}`, [["asOf", asOf]]);
  }

  async liveUpdatesByFixture(fixtureId: number): Promise<OddsPayload[]> {
    return await this.client.getJson(`/odds/updates/${fixtureId}`);
  }

  async historicalUpdates(options: {
    readonly epochDay: number;
    readonly hourOfDay: number;
    readonly interval: number;
    readonly fixtureId?: number;
  }): Promise<OddsPayload[]> {
    validateHour(options.hourOfDay);
    validateInterval(options.interval);
    return await this.client.getJson(
      `/odds/updates/${options.epochDay}/${options.hourOfDay}/${options.interval}`,
      [["fixtureId", options.fixtureId]],
    );
  }

  async validation(messageId: string, ts: number): Promise<OddsValidation> {
    return await this.client.getJson("/odds/validation", [
      ["messageId", messageId],
      ["ts", ts],
    ]);
  }
}
